### PR TITLE
Implement GET /variants/{id}

### PR DESCRIPTION
**Pull request by dcolligan**
Half-finished; will apply comments in #530 later

Issue #528

### CODE REVIEW

**Comment by dcolligan** [comment]
@jeromekelleher ready for review

File: ga4gh/datamodel/variants.py [hunk 1]
```diff
@@ -410,6 +445,26 @@ def convertVariant(self, record, callSetIds):
         variant.id = self.getVariantId(variant)
         return variant
 
+    def getVariant(self, compoundId):
+        if compoundId.referenceName in self._chromFileMap:
+            varFileName = self._chromFileMap[compoundId.referenceName]
+        else:
+            raise exceptions.ObjectNotFoundException(compoundId)
+        start = int(compoundId.start)
+        # fetch start is exclusive, not inclusive
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]: Does this comment still hold? You don't seem to be doing anything about it if it does...

File: ga4gh/datamodel/variants.py [hunk 2]
```diff
@@ -410,6 +445,26 @@ def convertVariant(self, record, callSetIds):
...
+        referenceName, startPosition, endPosition = \
+            self.sanitizeVariantFileFetch(
+                compoundId.referenceName, start, None)
+        cursor = self.getFileHandle(varFileName).fetch(
+            referenceName, startPosition, endPosition)
+        for record in cursor:
+            if start != record.start:
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]: What about
```python
variant = self.convertVariant(record, [])
if record.start == start and compoundId.md5 == self.hashVariant(variant):
    return variant
elif record.start > start:
    raise ObjectNotFoundException()
```
Most of the time, we'll hit the variant first go and so there's no point in trying to save a few cycles doing the convert variant process. This way we have only one method to get the hash. We do need to guard against the variant not existing, otherwise we could iterate over millions of variants before exiting the loop. Another way to do this I suppose is to put in an end coodinate. Is there any reason we can't use ``start + 1`` there?

File: tests/datadriven/test_variants.py [hunk 3]
```diff
@@ -346,3 +347,22 @@ def convertPyvcfNumber(number):
...
+        variantSet = self._gaObject
+        for referenceName in self._referenceNames:
+            refnameVariants = self._getPyvcfVariants(referenceName)
+            for variant in refnameVariants:
+                md5 = self._hashVariant(variant)
+                compoundId = datamodel.VariantCompoundId(
+                    variantSet.getCompoundId(), referenceName,
+                    variant.start, md5)
+                gotVariant = variantSet.getVariant(compoundId)
+                self.assertEqual(str(compoundId), gotVariant.id)
```
Comment (jeromekelleher) [inline tests/datadriven/test_variants.py]: We need some negative tests here --- what happens when we try to get some variants that are very like this one? Off by one coords, slightly different ref and alt?

**Comment by jeromekelleher** [comment]
Looks good @dcolligan, thanks. We need to think about what happens a bit more when variants are missing, but otherwise this is spot on.

**Comment by dcolligan** [comment]
@jeromekelleher pushed an update

**Comment by jeromekelleher** [comment]
Looks @dcolligan, thanks. 

+1 once the CI has completed (it failed with what looked like a pip error, so I kicked it off again).

**Comment by dcolligan** [comment]
no that was my bad, I resubmitted with a corrected branch

**Comment by jeromekelleher** [comment]
OK, merging. Thanks. That list of unimplemented URLS is getting small!

**Comment by dcolligan** [comment]
I want to get it down to 0, but that's assigned to someone else!
